cpp-terminal/base.hpp

```diff
@@ -1,109 +1,238 @@
-    yellow = 33,
-    blue = 34,
-    magenta = 35,
-    cyan = 36,
-    white = 37,
-    reset = 39,
-    gray = 90,
-    bright_red = 91,
-    bright_green = 92,
-    bright_yellow = 93,
-    bright_blue = 94,
-    bright_magenta = 95,
-    bright_cyan = 96,
-    bright_white = 97
+enum class Style : uint8_t {
+    RESET = 0,
+    BOLD = 1,
+    DIM = 2,
+    ITALIC = 3,
+    UNDERLINE = 4,
+    BLINK = 5,
+    BLINK_RAPID = 6,
+    REVERSED = 7,
+    CONCEAL = 8,
+    CROSSED = 9,
+    OVERLINE = 53
 };
 
-enum class bg {
-    black = 40,
-    red = 41,
-    green = 42,
-    yellow = 43,
-    blue = 44,
-    magenta = 45,
-    cyan = 46,
-    white = 47,
-    reset = 49,
-    gray = 100,
-    bright_red = 101,
-    bright_green = 102,
-    bright_yellow = 103,
-    bright_blue = 104,
-    bright_magenta = 105,
-    bright_cyan = 106,
-    bright_white = 107
+// Represents a RGB (24bit) color
+struct RGB {
+    uint8_t r{}, g{}, b{};
+    bool empty = false;
+};
+// indicates the color mode (basically the original color resolution)
+// also used to manually override the original color resolution
+enum class Mode {
+    // no color was used
+    NONE,
+    // a 4bit color was used
+    BIT4,
+    // a 8bit color was used
+    BIT8,
+    // a 24bit (RGB) color was used
+    BIT24,
+    // use 24bit colors, but fall back to 8bit if unsupported
+    AUTO24
+};
+// represents an RGB foreground and background color
+struct RGBF {
+    RGB rgb_fg;
+    Mode mode_fg{};
+    RGB rgb_bg;
+    Mode mode_bg{};
+};
+/*
+ * reference colors for converting RGB colors to 4bit colors and vice versa
+ */
+struct Bit4_reference {
+    static constexpr RGB BLACK{0, 0, 0, false};
+    static constexpr RGB RED{151, 12, 40, false};
+    static constexpr RGB GREEN{1, 142, 66, false};
+    static constexpr RGB YELLOW{238, 198, 67, false};
+    static constexpr RGB BLUE{13, 33, 161, false};
+    static constexpr RGB MAGENTA{255, 0, 144, false};
+    static constexpr RGB CYAN{0, 159, 184, false};
+    static constexpr RGB WHITE{240, 240, 240, false};
+    static constexpr RGB GRAY{127, 127, 127, false};
+    static constexpr RGB RED_BRIGHT{241, 85, 116, false};
+    static constexpr RGB GREEN_BRIGHT{52, 254, 146, false};
+    static constexpr RGB YELLOW_BRIGHT{243, 215, 124, false};
+    static constexpr RGB BLUE_BRIGHT{63, 136, 197, false};
+    static constexpr RGB MAGENTA_BRIGHT{255, 92, 184, false};
+    static constexpr RGB CYAN_BRIGHT{51, 228, 255, false};
+    static constexpr RGB WHITE_BRIGHT{255, 255, 255, false};
+    static constexpr RGB NONE{0, 0, 0, true};
 };
 
-std::string color(style);
-std::string color(fg);
-std::string color(bg);
-
-std::string color24_fg(unsigned int, unsigned int, unsigned int);
-
-std::string color24_bg(unsigned int, unsigned int, unsigned int);
-
-void write(const std::string&);
-
+// Converts a 4bit color to Term::RGB
+RGB bit4_to_rgb(Color4 color);
+// Converts a 8bit color to Term::RGB
+RGB bit8_to_rgb(uint8_t color);
```

Comment:
If I remember well `uint*_t` should be in `std` namespace when using `cstdint`. Maybe this could trigger somr warnings or compiler error on purist ones

Reply:
Sure, I'll change that.

Reply:
Implemented in 56056add80df701cf822502a2fcaa5c62cc9afb6